services/m3ctl/main/main.go

```diff
@@ -0,0 +1,155 @@
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/m3db/m3cluster/client/etcd"
+	"github.com/m3db/m3x/config"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+
+	"github.com/m3db/m3ctl/server"
+)
+
+type serverConfig struct {
+	// ListenAddress is the HTTP server listening address.
+	ListenAddress string `yaml:"listenAddress" validate:"nonzero"`
+
+	// ReadTimeout is the HTTP server read timeout.
+	ReadTimeout time.Duration `yaml:"readTimeout"`
+
+	// WriteTimeout HTTP server write timeout.
+	WriteTimeout time.Duration `yaml:"writeTimeout"`
+}
+
+// NewHTTPServerOptions create a new set of http server options.
+func (c *serverConfig) NewHTTPServerOptions(
+	instrumentOpts instrument.Options,
+) server.Options {
+	opts := server.NewOptions().SetInstrumentOptions(instrumentOpts)
+	if c.ReadTimeout != 0 {
+		opts = opts.SetReadTimeout(c.ReadTimeout)
+	}
+	if c.WriteTimeout != 0 {
+		opts = opts.SetWriteTimeout(c.WriteTimeout)
+	}
+
+	return opts
+}
+
+type kvClientConfiguration struct {
+	Etcd *etcd.Configuration `yaml:"etcd"`
+}
+
+type configuration struct {
+	// KVStore Config
+	KVClient kvClientConfiguration `yaml:"kvClient" validate:"nonzero"`
+
+	// Logging configuration.
+	Logging xlog.Configuration `yaml:"logging"`
+
+	// Metrics configuration.
+	Metrics instrument.MetricsConfiguration `yaml:"metrics"`
+
+	// HTTP server configuration.
+	HTTP serverConfig `yaml:"http"`
+}
+
+func main() {
+	configFile := flag.String("f", "", "configuration file")
+	flag.Parse()
+
+	if len(*configFile) == 0 {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	var cfg configuration
+	if err := xconfig.LoadFile(&cfg, *configFile); err != nil {
+		fmt.Printf("error loading config file: %v\n", err)
+		os.Exit(1)
+	}
+
+	logger, err := cfg.Logging.BuildLogger()
+	if err != nil {
+		fmt.Printf("error creating logger: %v\n", err)
+		os.Exit(1)
+	}
+
+	scope, closer, err := cfg.Metrics.NewRootScope()
+	if err != nil {
+		logger.Fatalf("error creating metrics root scope: %v", err)
+	}
+	defer closer.Close()
+
+	instrumentOpts := instrument.NewOptions().
+		SetLogger(logger).
+		SetMetricsScope(scope).
+		SetMetricsSamplingRate(cfg.Metrics.SampleRate()).
+		SetReportInterval(cfg.Metrics.ReportInterval())
+
+	listenAddr := cfg.HTTP.ListenAddress
+	serverOpts := cfg.HTTP.NewHTTPServerOptions(instrumentOpts)
+
+	doneCh := make(chan struct{})
+	closedCh := make(chan struct{})
+
+	s, err := server.NewServer(listenAddr, serverOpts)
+	if err != nil {
+		logger.Fatalf("could not create m3ctl server: %v", err)
+	}
+
+	go func() {
+		go func() {
+			logger.Infof("Starting http server on: %s", listenAddr)
+			if err = s.ListenAndServe(); err != nil {
+				logger.Fatalf("could not start serving traffic: %v", err)
+			}
+		}()
+		<-doneCh
+
+		logger.Debug("server closed")
+		close(closedCh)
+	}()
+
+	// Handle interrupts.
+	logger.Warnf("interrupt: %v", interrupt())
+
+	close(doneCh)
+
+	select {
+	case <-closedCh:
```

Comment:
This unfortunately doesn't do what you want, as soon as `doneCh` is closed, the goroutine will proceed to close `closedCh` with no knowledge about whether the server is actually closed.